### PR TITLE
feat(*): Add AGENTS.md file and custom agents

### PR DESCRIPTION
**Pull request by georgianastasov**
Closes: #17054

## Summary

This PR adds a new `AGENTS.md` file and a set of internal custom agents to guide AI-assisted work in the repository.

It also updates `.github/copilot-instructions.md` so the default Copilot guidance is better aligned with the repository structure.

## Why

The goal of these changes is to make AI-assisted contributions more consistent, more repo-aware, and easier to guide through a structured workflow.

### Updated

- `.github/copilot-instructions.md`
  - Refined the default Copilot guidance for the repository.
  - Keeps the instructions focused on modern Angular patterns, repository conventions, and existing project practices.

### Added

- `AGENTS.md`
  - Adds repository-wide guidance for AI agents.
  - Explains the project structure, working rules, commands, skills, and the role of each custom agent.
  - Helps keep AI-generated changes aligned with the repo’s conventions and expected workflow.

- `.github/README.md`
  - Adds documentation for the GitHub-side AI workflow files and related guidance.
  - Helps make the `.github/` setup easier to understand and navigate.

- `.github/agents/feature-orchestrator-agent.md`
  - Defines the top-level orchestration agent.
  - Used to analyze a request, plan the implementation, and delegate work to the specialized agents in sequence.
  - Helps keep feature work structured and consistent.
  
- `.github/agents/bug-fixing-orchestrator-agent.md`
  - Defines the top-level orchestration agent for bug fixing.
  - Used to investigate bug reports, identify scope and root cause, and delegate the bug-fix workflow to the appropriate specialized agents.
  - Helps keep bug fixing structured, reproducible, and aligned with the repository’s TDD and validation practices.

- `.github/agents/tdd-test-writer-agent.md`
  - Defines the test-writing agent for the RED phase of TDD.
  - Used to write failing tests before production code is added.
  - Helps enforce a more disciplined test-first workflow.
  
- `.github/agents/feature-implementer-agent.md`
  - Defines the implementation agent.
  - Used to write production code, make failing tests pass, and refactor where needed.
  - Helps keep feature implementation focused on minimal correct changes and repository conventions.
  
- `.github/agents/bug-fixing-implementer-agent.md`
  - Defines the implementation agent for bug fixing.
  - Used to apply the minimum production code fix needed to make a bug reproduction test pass while preserving existing public API and behavior whenever possible.
  - Helps keep bug fixes focused, low-risk, and consistent with repository conventions for accessibility, localization, and validation.

- `.github/agents/demo-sample-agent.md`
  - Defines the demo/sample update agent.
  - Used to update existing demo/sample areas under `src/app/` when a demo is explicitly requested for a user-visible feature or bug fix.
  - Helps keep demos aligned with the real implementation while reusing the existing sample structure.

- `.github/agents/theming-styles-agent.md`
  - Defines the theming and styles agent.
  - Used to implement SCSS, theme wiring, and style-validation changes in the in-repo styling system.
  - Helps keep visual changes aligned with the repository’s theming architecture and shared style rules.
  
- `.github/agents/component-readme-agent.md`
  - Defines the component README update agent.
  - Used to update affected component `README.md` files when public API or documented behavior changes.
  - Helps keep component documentation aligned with the actual implementation.

- `.github/agents/migration-agent.md`
  - Defines the migration agent.
  - Used to create `ng update` migration schematics when a breaking change is introduced.
  - Helps make breaking changes safer and easier to adopt.

- `.github/agents/changelog-agent.md`
  - Defines the changelog update agent.
  - Used to add changelog entries that follow the repository’s existing conventions.
  - Helps make release notes more consistent and easier to maintain.

## Result

Together, these files introduce a more structured internal AI workflow for the repository across both feature delivery and bug fixing:

- analyze and route the work with the appropriate orchestrator
- write failing tests first
- implement the feature or bug fix
- handle theming and styling work through a dedicated agent when needed
- update demo/sample areas when explicitly requested
- update component README files when needed
- add migrations when needed
- update the changelog when appropriate
- perform focused self-validation in each agent

This should make AI-assisted development more predictable, more maintainable, and more aligned with the repository’s existing standards.

## Notes

- These changes focus on internal AI guidance and workflow setup.

### CODE REVIEW

File: AGENTS.md [hunk 1]
```diff
@@ -0,0 +1,118 @@
+# AGENTS.md — Ignite UI for Angular
+
+This file defines repository-wide guidance for AI agents working in Ignite UI for Angular.
+
+## Project Overview
+
+Ignite UI for Angular is a comprehensive UI component library built on Angular 21+. It provides 60+ components including grids, charts, form controls, layout containers, and data visualization widgets.
```
Comment (kdinev) [inline AGENTS.md]:
```suggestion
Ignite UI for Angular is a comprehensive UI component library built on the Angular framework.
```

File: .github/agents/bug-fixing-implementer-agent.md [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+Treat failing tests as guidance, not as the full specification.
+
+---
+
+## How You Work
+
+1. **Read the original bug report** — understand expected vs. actual behavior.
+2. **Read the existing component source** — understand the current implementation, patterns, and conventions.
+3. **Read the failing test** — understand what behavior it is trying to reproduce.
```
Comment (kdinev) [inline .github/agents/bug-fixing-implementer-agent.md]: The skill references a failing reproduction test, which nothing told it to write. Consider either instructing the agent to write it, or making it `userinvocable: false`

File: .github/agents/bug-fixing-implementer-agent.md [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+---
+
+## Component-Specific Patterns
+
+Check the relevant skill file for component APIs and patterns:
+- Non-grid components → `skills/igniteui-angular-components/SKILL.md`
+- Grid components → `skills/igniteui-angular-grids/SKILL.md`
+- Theming & styling → `skills/igniteui-angular-theming/SKILL.md`
+
+Each skill file is a routing hub pointing to detailed reference files under its `references/` folder. **Read the relevant reference files in full** before modifying any component code.
```
Comment (kdinev) [inline .github/agents/bug-fixing-implementer-agent.md]: This is not needed, it knows it by default

File: .github/agents/bug-fixing-orchestrator-agent.md [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+
+## What You Do
+
+1. **Investigate the bug** — understand reproduction steps, expected vs. actual behavior
+2. **Locate affected code** — find relevant components, services, and files
+3. **Root-cause analysis** — identify the root cause before routing any work
+4. **Map impact** — determine what follow-through work is needed (migration, changelog, README, multi-branch)
+5. **Route work** — hand off to the right agents in the right order
+6. **Handle edge cases** — escalate when the bug cannot be reproduced, is by design, or is a third-party issue
+7. **Verify completeness** — check that nothing was missed after agents finish
```
Comment (kdinev) [inline .github/agents/bug-fixing-orchestrator-agent.md:64]: All of this seems to be covered already by the `bug-fixing-implementer`

File: .github/agents/bug-fixing-orchestrator-agent.md [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+Keep handoff framing minimal:
+- reference the original bug report
+- identify affected components or files
+- share the root cause finding
+- note whether migration, i18n, accessibility, or changelog follow-through may apply
+
+Do not restate the bug as:
+- detailed fix requirements
+- exact test cases
+- exact implementation instructions
```
Comment (kdinev) [inline .github/agents/bug-fixing-orchestrator-agent.md:87]: OK, this may actually cover my previous comment.

File: .github/agents/bug-fixing-orchestrator-agent.md [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+```
+
+---
+
+## Workflow
+
+### Step 1 — Investigate and Root-Cause
+
+1. Read the bug report carefully.
+2. Identify reproduction steps and expected vs. actual behavior.
```
Comment (kdinev) [inline .github/agents/bug-fixing-orchestrator-agent.md:144]: Maybe add a step to request additional context if required information is missing.

File: .github/agents/bug-fixing-orchestrator-agent.md [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+2. **`bug-fixing-implementer-agent`** — implements the minimum fix
+3. **`component-readme-agent`** — only if public API or documented behavior changed
+4. **`migration-agent`** — only if the fix introduces a breaking change
+5. **`changelog-agent`** — only if the fix warrants an entry under an existing CHANGELOG sections; otherwise leave `CHANGELOG.md` unchanged
+
+### Step 4 — Verify Completeness
+
+After all agents finish, check:
+
+- Does the failing test now pass?
```
Comment (kdinev) [inline .github/agents/bug-fixing-orchestrator-agent.md]:
```suggestion
- Does the newly added failing test now pass?
```

File: .github/agents/bug-fixing-orchestrator-agent.md [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+| By design / not a bug | Set `status: by-design` or `status: not a bug`. Comment referencing the relevant API documentation. Do not route to subagents. |
+| Third-party issue | Set `status: third-party-issue`. Document the external cause in the issue. Do not route to subagents. |
+| Fix requires breaking change | Route to `migration-agent` after the fix is implemented. Ensure CHANGELOG entry includes `BREAKING CHANGE`. |
+| Unrelated test failures | Note in the PR description. Do not attempt to fix unrelated failures. |
+
+---
+
+## Multi-branch Fixes
+
+When a bug exists in multiple release branches:
```
Comment (kdinev) [inline .github/agents/bug-fixing-orchestrator-agent.md]:
```suggestion
When a bug exists in multiple supported release branches:
```

File: .github/agents/bug-fixing-orchestrator-agent.md [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+1. Target the fix at the **oldest affected branch** first.
```
Comment (kdinev) [inline .github/agents/bug-fixing-orchestrator-agent.md]: Tell the agent how to extract what the currently supported branches are.

File: AGENTS.md [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+
+- Reuse existing repository patterns before introducing new abstractions.
+- Read existing source and tests before editing.
+- Run `npm run lint:lib` and the relevant test suite before considering work complete.
+- Export new public symbols from `<component>/index.ts`.
+- Keep accessibility intact: ARIA, keyboard navigation, focus behavior, and screen reader semantics.
+- Keep i18n intact when user-facing text changes.
+- Add JSDoc with `@param`, `@returns`, and `@example` on every new public member.
+- Add or update `ng update` migrations for true breaking changes such as removals, renames, moved entry points, selector changes, or incompatible default-behavior changes.
+- Update the component `README.md` when the public API surface changes.
```
Comment (kdinev) [inline AGENTS.md:58]:
```suggestion
- Update the component `README.md` when the public API surface changes.
- Update relevant Agent skills if a change is significant and/or you need to tell other agents how to use the newly introduced feature

```

File: .github/agents/tdd-test-writer-agent.md [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+---
+
+## How You Work
+
+1. **Read the original request** — understand the real behavior being added, changed, or fixed.
+2. **Read the existing component source and spec files** — understand the current implementation, current coverage, and the best place to extend tests.
+3. **Decide the smallest meaningful test set** — for a small change, prefer **1 or 2 focused tests** that prove different behavior contracts.
+4. **Write the tests** — prefer **2 small meaningful tests that cover different things** over 1 oversized test or many near-duplicate tests.
+5. Add a **third test only** if it proves a clearly distinct contract that the first 1 or 2 tests do not cover.
+6. **Run the tests** — confirm they fail for the intended missing behavior.
```
Comment (kdinev) [inline .github/agents/tdd-test-writer-agent.md]:
```suggestion
6. **Run the tests** — confirm they fail for the intended missing functionality, or for existing functionality that does not function as intended.
```

File: .github/agents/tdd-test-writer-agent.md [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+
+---
+
+## How You Work
+
+1. **Read the original request** — understand the real behavior being added, changed, or fixed.
+2. **Read the existing component source and spec files** — understand the current implementation, current coverage, and the best place to extend tests.
+3. **Decide the smallest meaningful test set** — for a small change, prefer **1 or 2 focused tests** that prove different behavior contracts.
+4. **Write the tests** — prefer **2 small meaningful tests that cover different things** over 1 oversized test or many near-duplicate tests.
+5. Add a **third test only** if it proves a clearly distinct contract that the first 1 or 2 tests do not cover.
```
Comment (kdinev) [inline .github/agents/tdd-test-writer-agent.md]: These instructions only apply to bug fixing. When doing feature development, the agent may want to plan a larger test suite by initial broader design of the feature and then to iteratively implement the feature and iteratively make the tests pass 1 by 1 or in batches related to the implemented functionality.

File: .github/agents/tdd-test-writer-agent.md [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+
+---
+
+## Rules
+
+1. **Default to the smallest useful test set.**
+   - For a small additive change or bug fix, prefer **1 or 2 focused tests**.
+   - Prefer **2 focused tests that cover different behaviors** over 1 oversized test.
+   - Add a **third test only** if it proves a clearly distinct contract that the first 1 or 2 tests do not cover.
+   - Do **not** write more than **3 new tests total** unless the user explicitly asks for broader coverage.
```
Comment (kdinev) [inline .github/agents/tdd-test-writer-agent.md]: Again the same comment as above.

File: .github/agents/tdd-test-writer-agent.md [hunk 14]
```diff
@@ -0,0 +1,174 @@
...
+   - Use the most relevant existing `describe` block whenever it is reasonably suitable.
+   - Create a new `describe` block only if reusing an existing one would be misleading or confusing.
+4. **Tests must fail for the intended new behavior.**
+   - If a test passes immediately, it is not testing new behavior. Fix it.
+   - If a test fails only because of unrelated setup or missing symbols, fix the test.
+5. **Tests must be independent.**
+   - No shared mutable state.
+   - No execution-order dependency.
+6. **Never write production code.**
+   - Only test code in this phase.
```
Comment (kdinev) [inline .github/agents/tdd-test-writer-agent.md]:
```suggestion
   - Only test code in this phase to ensure the test does not give false negatives. The test MUST fail if the functionality is missing, or not behaving as intended.
```

File: .github/agents/migration-agent.md [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+
+# Migration Schematic Agent
+
+You create **`ng update` migration schematics** for breaking changes in Ignite UI for Angular. Every breaking change (removed API, renamed type/property/enum, moved entry point, changed default behavior) **must** have a migration so consumer apps update automatically.
+
+---
+
+## When to Create Migrations
+
+- API removed or renamed
```
Comment (kdinev) [inline .github/agents/migration-agent.md]:
```suggestion
- API removed, renamed, or deprecated
```

File: .github/agents/feature-implementer-agent.md [hunk 16]
```diff
@@ -0,0 +1,136 @@
...
+2. **Test code**: extract shared setup and sharpen assertions. Add or adjust tests only when required by the actual feature contract.
+3. Run tests — confirm no regressions.
+
+---
+
+## API Documentation
+
+Add JSDoc on every new or changed public member with `@param`, `@returns`, and `@example`.
+
+---
```
Comment (kdinev) [inline .github/agents/feature-implementer-agent.md:107]:
```suggestion
---

## Agent Skills

Update component agent skills if you need to guide other agents on how to use the newly added feature.

---
```

File: .github/agents/bug-fixing-orchestrator-agent.md (hunk 4 above)
Comment (ivanvpetrov) [reply .github/agents/bug-fixing-orchestrator-agent.md:64]: You're comment got me thinking that both 

- the orchestrator has to do a bug analysis so that it can guide and analyze the work of the subagents
- the implementor has to do a bug analysis to fix the bug, especially if called as a standalone agent (since it's user invokable)

Which may lead to repetitive work by the agents, token wastage and not clear separation of concerns.

Thant's why I decided to introduce a '**Bug Knowledge Block**' that get's initially filled by the orchestrator's investigation of the bug. Then this block is explicitly passed to the sub agents, so that they do not perform an investigation of their own again to do their job. 

Also, the bug-fixing-implementor now has two modes. To operate with an provided bug knowledge block (orchestrated) or to do it's own investigation (stand-alone) 

[bug knowledge block commit](https://github.com/IgniteUI/igniteui-angular/pull/17053/changes/21362327c441d5bc35c5ff1eab2da9ad3ab151e1)

File: .github/agents/bug-fixing-implementer-agent.md (hunk 2 above)
Comment (ivanvpetrov) [reply .github/agents/bug-fixing-implementer-agent.md]: Now, once the tdd-agent creates or identifies the failing test, it get's added as reference to the **Bug Knowledge block** (see the comment further down about this block) and is passed down the stream to the bug-fixing-implementer agent. So now there is no ambiguity which is this failing test.

File: .github/agents/bug-fixing-orchestrator-agent.md (hunk 4 above)
Comment (ivanvpetrov) [reply .github/agents/bug-fixing-orchestrator-agent.md:64]: P.S. 
If we still want concrete separation of concerns, we should introduce an additional **bug-investigation-agent** (probably user none-invokable), which only investigate the bug and creates this **Bug Knowledge block**

That way:
- The orchestrator will call it as a first step
- The bug-fixing-implementor will call it **if working in stand-alone mode**

It will be purer from an architecture point of view, but probably an overkill.

File: .github/agents/bug-fixing-orchestrator-agent.md (hunk 9 above)
Comment (ivanvpetrov) [reply .github/agents/bug-fixing-orchestrator-agent.md]: Additional instructions for identifying supported versions are added. Lookup in SECURITY.md 
```yaml
Only create PRs for **supported versions**. Before acting, read [SECURITY.md](../../SECURITY.md) to get the current list of supported versions. Do **not** create PRs for any version marked as unsupported there.
....
1. Read SECURITY.md and collect all versions currently marked as supported.
2. Determine which of those supported branches are actually affected — a bug introduced in a later version does not need a fix in earlier branches.
```

File: .github/README.md [hunk 17]
```diff
@@ -0,0 +1,400 @@
...
+- implementation follows
+- README, migration, and changelog work are **optional follow-through**
+
+---
+
+## Agent List
+
+| Agent | Role | Use it for |
+|---|---|---|
+| [`feature-orchestrator-agent`](./agents/feature-orchestrator-agent.md) | Orchestrator | New features, enhancements, broader feature work |
```
[the remark this answers is not in the document]
Comment (kdinev) [reply .github/README.md:110]: @copilot open a new pull request to apply changes based on [this feedback](https://github.com/IgniteUI/igniteui-angular/pull/17053#discussion_r2973983312)

File: skills/igniteui-angular-theming/references/contributing.md [hunk 18]
```diff
@@ -0,0 +1,471 @@
...
+```scss
+@use '../../base' as *;
+```
+
+---
+
+## The Two-File Pattern
+
+Every component has exactly two SCSS files. **Never merge them.**
+
```
[the remark this answers is not in the document]
Comment (kdinev) [reply skills/igniteui-angular-theming/references/contributing.md:48]: @simeonoff
